### PR TITLE
(maint) Fix segfault due to invalid SID in tests

### PR DESCRIPTION
**Pull request by joshcooper**
Calling `SID.octet_string_to_principal` with an invalid byte array containing a
single element `2` could cause a segfault on Windows because
`Principal#lookup_account_sid` converts the byte array to a SID pointer (PSID)
and calls `LookupAccountSidW` with it. Apparently, Windows doesn't sanity check
the argument before using it, sometimes resulting in the following:

    C:/Users/josh/projects/puppet/lib/puppet/util/windows/principal.rb:115: [BUG] Segmentation fault
    ruby 2.5.7p206 (2019-10-01 revision 67816) [x64-mingw32]

    -- Control frame information -----------------------------------------------
    c:0062 p:---- s:0321 e:000320 CFUNC  :LookupAccountSidW
    c:0061 p:0050 s:0310 e:000309 BLOCK  C:/Users/josh/projects/puppet/lib/puppet/util/windows/principal.rb:115 [FINISH]
    c:0060 p:---- s:0304 e:000303 CFUNC  :initialize
    c:0059 p:---- s:0301 e:000300 CFUNC  :new
    c:0058 p:0017 s:0295 e:000294 BLOCK  C:/Users/josh/projects/puppet/lib/puppet/util/windows/principal.rb:112 [FINISH]
    c:0057 p:---- s:0291 e:000290 CFUNC  :initialize
    c:0056 p:---- s:0288 e:000287 CFUNC  :new
    c:0055 p:0017 s:0282 e:000281 BLOCK  C:/Users/josh/projects/puppet/lib/puppet/util/windows/principal.rb:111 [FINISH]
    c:0054 p:---- s:0278 e:000277 CFUNC  :initialize
    c:0053 p:---- s:0275 e:000274 CFUNC  :new
    c:0052 p:0017 s:0269 e:000268 BLOCK  C:/Users/josh/projects/puppet/lib/puppet/util/windows/principal.rb:110 [FINISH]
    c:0051 p:---- s:0265 e:000264 CFUNC  :initialize
    c:0050 p:---- s:0262 e:000261 CFUNC  :new
    c:0049 p:0158 s:0256 e:000255 METHOD C:/Users/josh/projects/puppet/lib/puppet/util/windows/principal.rb:109
    c:0048 p:0065 s:0248 e:000247 METHOD C:/Users/josh/projects/puppet/lib/puppet/util/windows/sid.rb:97
    c:0047 p:0014 s:0243 e:000242 BLOCK  C:/Users/josh/projects/puppet/spec/unit/util/windows/sid_spec.rb:54 [FINISH]

Verify the PSID before trying to use it, and return `ERROR_INVALID_PARAMETER`
like we did before.

### CODE REVIEW

File: lib/puppet/util/windows/principal.rb [hunk 1]
```diff
@@ -112,6 +113,11 @@ def self.lookup_account_sid(system_name = nil, sid_bytes)
               FFI::MemoryPointer.new(:uint32, 1) do |name_use_enum_ptr|
 
                 sid_ptr.write_array_of_uchar(sid_bytes)
+
+                if Puppet::Util::Windows::SID.IsValidSid(sid_ptr) == FFI::WIN32_FALSE
+                  raise Puppet::Util::Windows::Error.new(_('Byte array for lookup_account_sid is invalid: %{sid_bytes}') % { sid_bytes: sid_bytes }, ERROR_INVALID_PARAMETER)
```
Comment (joshcooper) [inline lib/puppet/util/windows/principal.rb:118]: Probably shouldn’t print the binary array. Or hex encode it?

Comment (joshcooper) [reply lib/puppet/util/windows/principal.rb:118]: Similar problem in https://github.com/puppetlabs/puppet/pull/8202/files/54a14b291bb1f749961a5a161fba7efc33128556#diff-88ba5e4691d33bdb2fd674511129fa6fR126

and https://github.com/puppetlabs/puppet/pull/8202/files/54a14b291bb1f749961a5a161fba7efc33128556#diff-88ba5e4691d33bdb2fd674511129fa6fR133

Comment (joshcooper) [reply lib/puppet/util/windows/principal.rb:118]: This actually works ok: 

```
irb(main):003:0> Puppet::Util::Windows::SID.octet_string_to_principal([2, 3, 4, 5, 6])
Puppet::Util::Windows::Error: Byte array for lookup_account_sid is invalid: [2, 3, 4, 5, 6]:  The parameter is incorrect.
```

**Comment by Iristyle** [comment]
You finally found it? Hurray!

**Pull request merged**